### PR TITLE
gnrc_sock_udp: choose random ephemeral port

### PR DESCRIPTION
**Pull request by nmeum**
### Contribution description

While writing a PoC for #12293 and #12382 I noticed that RIOT does not randomize its ephemeral currently. This circumstance also made it significantly easier to exploit #10739. The first port  value is `49152` and the value is simply sequential incremented after that. I consider this a security issue as it eases spoofing replies why this is a problem is hopefully illustrated by the previously mentioned issues.

While working on this change, I noticed that `gnrc_sock_udp` already depends on the `random` module in `Makefile.dep`, even though it does not use it. This made me curious and I discovered that it was removed, without much further discussion, in #6212: 

> I was kind of proud to get rid of the random() …

I personally think that this is a bad idea and an unnecessary security risk. If there is any specific reason why I should not be randomized please let me know.

### Testing procedure

I tested this change using `examples/asymcute_mqttsn/` simply build the application, start it. Try to connect to a non-existing MQTT server and check using wireshark/tcpdump that it uses a random port as the UDP source port.

### Related

* #12293 
* #12382
* #10739
* #6212

### CODE REVIEW

**Comment by smlng** [comment]
actually there was a [discussion](https://github.com/RIOT-OS/RIOT/pull/6212/files#r92625194) regarding random or not between @OlegHahm and me (@smlng), the reasoning was to not introduce the additional overhead of random by default. However, we also wanted to discuss this (or the implementation according to RFC 6056) separately of the PR #6212 - which sadly never happened. 

Maybe this PR takes this up now -

**Comment by smlng** [comment]
on another note: (to me) in the past we (RIOT developers) were more into runtime and binary optimisation back when, but today I generally agree that features and security (should) come first. So I'm in favour of (re)introducing random here, but please review [RFC 6056](https://tools.ietf.org/html/rfc6056) too.

**Comment by nmeum** [comment]
> actually there was a discussion regarding random or not between @OlegHahm and me (@smlng), the reasoning was to not introduce the additional overhead of random by default.

Thanks for pointing that out, I somehow missed that thread in the pull request, sorry.

> […] please review [RFC 6056](https://tools.ietf.org/html/rfc6056) too.

Didn't know about that RFC, thanks a lot for referencing it. In fact, the comment above `_get_dyn_port` function claims that the function implements the [Simple Hash-Based Port Selection Algorithm](https://tools.ietf.org/html/rfc6056#section-3.3.3) from this RFC. My original intention was to implement [Another Simple Port Randomization Algorithm](https://tools.ietf.org/html/rfc6056#section-3.3.2), though my version still multiplies with `GNRC_SOCK_DYN_PORTRANGE_OFF` by accident.

RFC 6056 has an interesting section on [Choosing an Ephemeral Port Selection Algorithm](https://tools.ietf.org/html/rfc6056#section-3.5). According to this section, the algorithm proposed here "increase the chances of port number collisions, which could lead to the failure of a connection establishment attempt.". However, in my experience constrained devices establish far less connections than conventional ones. For this reason, I deem collisions to be less likely and have a personal preference towards this algorithm as it chooses a new truly random port for each socket.

I will update the PR to make sure it implements [Another Simple Port Randomization Algorithm](https://tools.ietf.org/html/rfc6056#section-3.3.2) correctly. If you prefer a different algorithm let me know.

File: sys/net/gnrc/sock/udp/gnrc_sock_udp.c [hunk 1]
```diff
@@ -66,15 +65,15 @@ static bool _dyn_port_used(uint16_t port)
...
  */
 static uint16_t _get_dyn_port(sock_udp_t *sock)
 {
     unsigned count = GNRC_SOCK_DYN_PORTRANGE_NUM;
     do {
         uint16_t port = GNRC_SOCK_DYN_PORTRANGE_MIN +
-               (_dyn_port_next * GNRC_SOCK_DYN_PORTRANGE_OFF) % GNRC_SOCK_DYN_PORTRANGE_NUM;
-        _dyn_port_next++;
+               (random_uint32() % GNRC_SOCK_DYN_PORTRANGE_NUM);
         if ((sock == NULL) || (sock->flags & SOCK_FLAGS_REUSE_EP) ||
```
Comment (nmeum) [inline sys/net/gnrc/sock/udp/gnrc_sock_udp.c]: Does anybody have an idea why the `sock == NULL` check is necessary?

Comment (smlng) [reply sys/net/gnrc/sock/udp/gnrc_sock_udp.c]: IIRC the compiler complaints that `sock->flags` might not work if `sock` is used uninitialised. I think I was also arguing back then, that this is an internal function and hence can make sure sock is always not `NULL`.

Comment (smlng) [reply sys/net/gnrc/sock/udp/gnrc_sock_udp.c]: But the compiler (or static code checker??) wouldn't calm down without this check.

Comment (nmeum) [reply sys/net/gnrc/sock/udp/gnrc_sock_udp.c]: But in that case the check is incorrect as it should be:

```C
         if ((sock != NULL && (sock->flags & SOCK_FLAGS_REUSE_EP)) || 
             !_dyn_port_used(port))
```

then.

Comment (smlng) [reply sys/net/gnrc/sock/udp/gnrc_sock_udp.c]: mhm, you're right the '||' doesn't make sense here. Also I looked up all usages of `_get_dyn_port` and the calling function already makes sure (via `assert` or other) that `sock != NULL`. So I'd say you can safely remove that check, though you could add an `assert(sock)` if you want to be double safe.

Comment (nmeum) [reply sys/net/gnrc/sock/udp/gnrc_sock_udp.c]: Ok, good to know. Thanks for researching this. Adjusted the code accordingly.

**Comment by nmeum** [comment]
@smlng seems the assumption about `sock` never being null is incorrect for `tests_gnrc_sock_udp`:

```
main(): This is RIOT! (Version: buildtest)
Calling test_sock_udp_create__EADDRINUSE()
Calling test_sock_udp_create__EAFNOSUPPORT()
Calling test_sock_udp_create__EINVAL_addr()
Calling test_sock_udp_create__EINVAL_netif()
Calling test_sock_udp_create__no_endpoints()
Calling test_sock_udp_create__only_local()
Calling test_sock_udp_create__only_local_port0()
Calling test_sock_udp_create__only_local_reuse_ep()
Calling test_sock_udp_create__only_remote()
Calling test_sock_udp_create__full()
Calling test_sock_udp_recv__EADDRNOTAVAIL()
Calling test_sock_udp_recv__EAGAIN()
Calling test_sock_udp_recv__ENOBUFS()
Calling test_sock_udp_recv__EPROTO()
Calling test_sock_udp_recv__ETIMEDOUT()
 * Calling sock_udp_recv()
 * (timed out with timeout 1000000)
Calling test_sock_udp_recv__socketed()
Calling test_sock_udp_recv__socketed_with_remote()
Calling test_sock_udp_recv__socketed_with_port0()
Calling test_sock_udp_recv__unsocketed()
Calling test_sock_udp_recv__unsocketed_with_remote()
Calling test_sock_udp_recv__with_timeout()
Calling test_sock_udp_recv__non_blocking()
Calling test_sock_udp_send__EAFNOSUPPORT()
Calling test_sock_udp_send__EINVAL_addr()
Calling test_sock_udp_send__EINVAL_netif()
Calling test_sock_udp_send__EINVAL_port()
Calling test_sock_udp_send__ENOTCONN()
Calling test_sock_udp_send__socketed_no_local_no_netif()
Calling test_sock_udp_send__socketed_no_netif()
Calling test_sock_udp_send__socketed_no_local()
Calling test_sock_udp_send__socketed()
Calling test_sock_udp_send__socketed_other_remote()
Calling test_sock_udp_send__unsocketed_no_local_no_netif()
Calling test_sock_udp_send__unsocketed_no_netif()
Calling test_sock_udp_send__unsocketed_no_local()
Calling test_sock_udp_send__unsocketed()
Calling test_sock_udp_send__no_sock_no_netif()
sys/net/gnrc/sock/udp/gnrc_sock_udp.c:73 => 0x565eb1fb
*** RIOT kernel panic:
FAILED ASSERTION.

	pid | name                 | state    Q | pri | stack  ( used) | base addr  | current     
	  - | isr_stack            | -        - |   - |   8192 (   -1) | 0x56610580 | 0x56610580
	  1 | idle                 | pending  Q |  15 |   8192 (  568) | 0x5660e3a0 | 0x56610210 
	  2 | main                 | running  Q |   7 |  12288 ( 2992) | 0x5660b3a0 | 0x5660e210 
	  3 | ipv6                 | bl rx    _ |   4 |   8192 ( 1440) | 0x56619520 | 0x5661b390 
	  4 | udp                  | bl rx    _ |   5 |   8192 ( 1264) | 0x566174c0 | 0x56619330 
	    | SUM                  |            |     |  45056 ( 6264)

*** halted.
```

I changed the check to:

```C
if ((sock != NULL && (sock->flags & SOCK_FLAGS_REUSE_EP)) ||
    !_dyn_port_used(port))
```

**Comment by smlng** [comment]
so there was a reason for the check after all 😉 didn't remember that.

File: sys/net/gnrc/sock/udp/gnrc_sock_udp.c [hunk 2]
```diff
@@ -66,17 +65,16 @@ static bool _dyn_port_used(uint16_t port)
...
 {
     unsigned count = GNRC_SOCK_DYN_PORTRANGE_NUM;
     do {
         uint16_t port = GNRC_SOCK_DYN_PORTRANGE_MIN +
-               (_dyn_port_next * GNRC_SOCK_DYN_PORTRANGE_OFF) % GNRC_SOCK_DYN_PORTRANGE_NUM;
-        _dyn_port_next++;
-        if ((sock == NULL) || (sock->flags & SOCK_FLAGS_REUSE_EP) ||
-                              !_dyn_port_used(port)) {
+               (random_uint32() % GNRC_SOCK_DYN_PORTRANGE_NUM);
+        if ((sock && (sock->flags & SOCK_FLAGS_REUSE_EP)) || !_dyn_port_used(port)) {
```
Comment (smlng) [inline sys/net/gnrc/sock/udp/gnrc_sock_udp.c]: `&&`?  I'd suggest to leave as it was before, right?

Comment (nmeum) [reply sys/net/gnrc/sock/udp/gnrc_sock_udp.c]: I thought [we agreed](https://github.com/RIOT-OS/RIOT/pull/13253#discussion_r375118405) that the previous check was broken:

> mhm, you're right the '||' doesn't make sense here.

The previous check would not check for local port collisions if `sock` is NULL, which seems very wrong to me.

**Comment by miri64** [comment]
> @smlng seems the assumption about `sock` never being null is incorrect for `tests_gnrc_sock_udp`:

Yepp, very useful for simple fire-and-forget solutions ;-)

File: sys/net/gnrc/sock/udp/gnrc_sock_udp.c (hunk 2 above)
Comment (nmeum) [reply sys/net/gnrc/sock/udp/gnrc_sock_udp.c]: ?

**Comment by nmeum** [comment]
Any reason why this hasn't been merged yet? IMHO this change is straightforward. I would also like to point out that [RFC 6056](https://tools.ietf.org/html/rfc6056) explicitly discourages using [Algorithm 3](https://tools.ietf.org/html/rfc6056#section-3.3.3) (which is **incorrectly** implemented at the moment anyhow as `_get_dyn_port` uses a constant instead of a hash function atm) without a cryptographic hash function with the proper arguments and advertises using [Algorithm 2](https://tools.ietf.org/html/rfc6056#section-3.3.2) then. Maybe someone else can take a look at this since @smlng hasn't responded lately. @miri64 maybe?

**Comment by miri64** [comment]
I suspect @smlng disappeared in children-assisted home office ;-) so let me have a look.

**Comment by miri64** [comment]
Last compile test was >1 month ago. Let's give that a rerun.

**Comment by miri64** [comment]
Failing tests are unrelated and also fail in the current nightlies.

**Pull request merged**